src/Features/Core/Portable/RemoveAsyncModifier/AbstractRemoveAsyncModifierCodeFixProvider.cs

```diff
@@ -0,0 +1,261 @@
+            var compilation = semanticModel.Compilation;
+            var knownTypes = new KnownTypes(compilation);
+
+            foreach (var diagnostic in diagnostics.OrderByDescending(d => d.Location.SourceSpan.Start))
+            {
+                var token = diagnostic.Location.FindToken(cancellationToken);
+                var node = token.GetAncestor(IsAsyncSupportingFunctionSyntax);
+                if (node == null)
+                {
+                    Debug.Fail("We should always be able to find the node from the diagnostic.");
+                    continue;
+                }
+
+                var methodSymbol = GetMethodSymbol(node, semanticModel, cancellationToken);
+                if (methodSymbol == null)
+                {
+                    Debug.Fail("We should always be able to find the method symbol for the diagnostic.");
+                    continue;
+                }
+
+                // We might need to perform control flow analysis as part of the fix, so we need to do it on the original node
+                // so do it up front. Nothing in the fixer changes the reachability of the end of the method so this is safe
+                var controlFlow = GetControlFlowAnalysis(generator, semanticModel, node);
+                // If control flow couldn't be computed then its probably an empty block, which means we need to add a return anyway
+                var needsReturnStatementAdded = controlFlow == null || controlFlow.EndPointIsReachable;
+
+                editor.ReplaceNode(node, (updatedNode, generator) => RemoveAsyncModifier(generator, updatedNode, methodSymbol.ReturnType, knownTypes, needsReturnStatementAdded));
+            }
+        }
+
+        private static IMethodSymbol? GetMethodSymbol(SyntaxNode node, SemanticModel semanticModel, CancellationToken cancellationToken)
+            => semanticModel.GetSymbolInfo(node, cancellationToken).Symbol is IMethodSymbol methodSymbol
+                ? methodSymbol
+                : semanticModel.GetDeclaredSymbol(node, cancellationToken) as IMethodSymbol;
+
+        private static bool ShouldOfferFix(ITypeSymbol returnType, KnownTypes knownTypes)
+            => IsTaskType(returnType, knownTypes)
+                || returnType.OriginalDefinition.Equals(knownTypes._taskOfTType)
+                || returnType.OriginalDefinition.Equals(knownTypes._valueTaskOfTTypeOpt);
+
+        private static bool IsTaskType(ITypeSymbol returnType, KnownTypes knownTypes)
+            => returnType.OriginalDefinition.Equals(knownTypes._taskType)
+                || returnType.OriginalDefinition.Equals(knownTypes._valueTaskType);
+
+        private SyntaxNode RemoveAsyncModifier(SyntaxGenerator generator, SyntaxNode node, ITypeSymbol returnType, KnownTypes knownTypes, bool needsReturnStatementAdded)
+        {
+            node = RemoveAsyncModifier(generator, node);
```

Comment:
nit: up to you, but my preference is to not overwrite in this fashion as it gets harder to reason about which node this is referring to as hte method progresses.

Reply:
Going to leave this for now just to avoid the rework it would require, but will keep it in mind for next time. I can see the benefit, but it also means coming up with lots more variable names, and at least by reusing `node` i'm sure each bit is operating on the latest version.